include/small/slab_cache_asan.h

```diff
@@ -0,0 +1,121 @@
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "quota_lessor.h"
+#include "util.h"
+
+#include <stddef.h>
+#include <stdint.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+enum {
+	/** Alignment for slabs from ASAN slab cache. */
+	SMALL_CACHE_SLAB_ALIGNMENT = sizeof(long long),
+};
+
+/**
+ * ASAN friendly implementation for slab cache allocator. It has same
+ * interface as regular implementation but allocates every allocation using
+ * malloc(). This allows to do usual ASAN checks for memory allocation.
+ *
+ * This implementation do not follow no regular implementation slab order
+ * nor slab alignment.
+ */
+struct slab_cache {
+	/**
+	 * Allocation quota. It is not used directly. We need to keep quota
+	 * to pass it to small object allocator only.
+	 */
+	struct quota_lessor quota;
+	size_t used;
+	/** Unique id among all allocators. */
+	unsigned int id;
+};
+
+/** Header of slab cache allocation. */
+struct slab {
+	/** Allocation size. */
+	size_t size;
+};
+
+/** Extra data associated with each slab. */
+struct slab_cache_object {
+	/** id of the cache that this object belongs to. */
+	unsigned int cache_id;
+};
+
+void
+slab_cache_create(struct slab_cache *cache, struct slab_arena *arena);
+
+static inline void
+slab_cache_destroy(struct slab_cache *cache)
+{
+	quota_lessor_destroy(&cache->quota);
+}
+
+static inline size_t
+slab_real_size(struct slab_cache *cache, size_t size)
+{
+	(void)cache;
+	/* Copy-paste of slab_sizeof() */
+	size_t meta_size = small_align(sizeof(struct slab), sizeof(intptr_t));
+	return small_align(size + meta_size, small_getpagesize());
+}
+
+struct slab *
+slab_get(struct slab_cache *cache, size_t size);
+
+void
+slab_put(struct slab_cache *cache, struct slab *slab);
+
+static inline size_t
+slab_cache_used(struct slab_cache *cache)
+{
+	return cache->used;
+}
+
+static inline void
+slab_cache_check(struct slab_cache *cache)
+{
+	(void)cache;
+}
+
+static inline void
+slab_cache_set_thread(struct slab_cache *cache)
+{
+	(void)cache;
```

Comment:
Why not implement the thread check for the asan version?

Reply:
Added check similar to non-ASAN version.

<details>
<summary>diff</summary>

```diff
diff --git a/include/small/slab_cache.h b/include/small/slab_cache.h
index b5e8aa6..078d86c 100644
--- a/include/small/slab_cache.h
+++ b/include/small/slab_cache.h
@@ -31,6 +31,7 @@
  * SUCH DAMAGE.
  */
 #include "slab_arena.h"
+#include <pthread.h>
 
 #include "small_config.h"
 
@@ -45,7 +46,6 @@
 #include <stddef.h>
 #include <assert.h>
 #include "rlist.h"
-#include <pthread.h>
 #include "slab_list.h"
 #include "util.h"
 
diff --git a/include/small/slab_cache_asan.h b/include/small/slab_cache_asan.h
index 7daaa2e..72c4acf 100644
--- a/include/small/slab_cache_asan.h
+++ b/include/small/slab_cache_asan.h
@@ -60,6 +60,9 @@ struct slab_cache {
 	size_t used;
 	/** Unique id among all allocators. */
 	unsigned int id;
+#ifndef NDEBUG
+	pthread_t thread_id;
+#endif
 };
 
 /** Header of slab cache allocation. */
@@ -114,6 +117,9 @@ static inline void
 slab_cache_set_thread(struct slab_cache *cache)
 {
 	(void)cache;
+#ifndef NDEBUG
+	cache->thread_id = pthread_self();
+#endif
 }
 
 #if defined(__cplusplus)
diff --git a/small/slab_cache_asan.c b/small/slab_cache_asan.c
index aa18190..31107d7 100644
--- a/small/slab_cache_asan.c
+++ b/small/slab_cache_asan.c
@@ -41,6 +41,7 @@ slab_cache_create(struct slab_cache *cache, struct slab_arena *arena)
 SMALL_NO_SANITIZE_ADDRESS struct slab *
 slab_get(struct slab_cache *cache, size_t size)
 {
+	assert(pthread_equal(cache->thread_id, pthread_self()));
 	size = slab_real_size(cache, size);
 	struct slab_cache_object *obj =
 			small_asan_alloc(size,
@@ -56,6 +57,7 @@ slab_get(struct slab_cache *cache, size_t size)
 SMALL_NO_SANITIZE_ADDRESS void
 slab_put(struct slab_cache *cache, struct slab *slab)
 {
+	assert(pthread_equal(cache->thread_id, pthread_self()));
 	struct slab_cache_object *obj = small_asan_header_from_payload(slab);
 	small_asan_assert(obj->cache_id == cache->id &&
 			  "object and cache id mismatch");

```

</details>

Reply:
Let's enable it for release builds as well using `slab_asan_assert`.

Reply:
<details>
<summary>new diff</summary>

```diff
diff --git a/include/small/slab_cache.h b/include/small/slab_cache.h
index b5e8aa6..078d86c 100644
--- a/include/small/slab_cache.h
+++ b/include/small/slab_cache.h
@@ -31,6 +31,7 @@
  * SUCH DAMAGE.
  */
 #include "slab_arena.h"
+#include <pthread.h>
 
 #include "small_config.h"
 
@@ -45,7 +46,6 @@
 #include <stddef.h>
 #include <assert.h>
 #include "rlist.h"
-#include <pthread.h>
 #include "slab_list.h"
 #include "util.h"
 
diff --git a/include/small/slab_cache_asan.h b/include/small/slab_cache_asan.h
index 7daaa2e..913b4f5 100644
--- a/include/small/slab_cache_asan.h
+++ b/include/small/slab_cache_asan.h
@@ -60,6 +60,7 @@ struct slab_cache {
 	size_t used;
 	/** Unique id among all allocators. */
 	unsigned int id;
+	pthread_t thread_id;
 };
 
 /** Header of slab cache allocation. */
@@ -114,6 +115,7 @@ static inline void
 slab_cache_set_thread(struct slab_cache *cache)
 {
 	(void)cache;
+	cache->thread_id = pthread_self();
 }
 
 #if defined(__cplusplus)
diff --git a/small/slab_cache_asan.c b/small/slab_cache_asan.c
index aa18190..af89995 100644
--- a/small/slab_cache_asan.c
+++ b/small/slab_cache_asan.c
@@ -36,11 +36,13 @@ slab_cache_create(struct slab_cache *cache, struct slab_arena *arena)
 {
 	quota_lessor_create(&cache->quota, arena->quota);
 	cache->id = small_asan_reserve_id();
+	slab_cache_set_thread(cache);
 }
 
 SMALL_NO_SANITIZE_ADDRESS struct slab *
 slab_get(struct slab_cache *cache, size_t size)
 {
+	small_asan_assert(pthread_equal(cache->thread_id, pthread_self()));
 	size = slab_real_size(cache, size);
 	struct slab_cache_object *obj =
 			small_asan_alloc(size,
@@ -56,6 +58,7 @@ slab_get(struct slab_cache *cache, size_t size)
 SMALL_NO_SANITIZE_ADDRESS void
 slab_put(struct slab_cache *cache, struct slab *slab)
 {
+	small_asan_assert(pthread_equal(cache->thread_id, pthread_self()));
 	struct slab_cache_object *obj = small_asan_header_from_payload(slab);
 	small_asan_assert(obj->cache_id == cache->id &&
 			  "object and cache id mismatch");
```

</details>